src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,299 @@
+@[simp] lemma coe_one : ⇑(1 : centroid_hom α) = id := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+@[simp] lemma coe_mul (f g : centroid_hom α) : ⇑(f * g) = f ∘ g := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma coe_nsmul (f : centroid_hom α) (n : ℕ) : ⇑(n • f) = n • f := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma one_apply (a : α) : (1 : centroid_hom α) a = a := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+@[simp] lemma mul_apply (f g : centroid_hom α) (a : α) : (f * g) a = f (g a) := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf]
+lemma nsmul_apply (f : centroid_hom α) (n : ℕ) (a : α) : (n • f) a = n • f a := rfl
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance : has_nat_cast (centroid_hom α) :=
+{ nat_cast := λ n, n • 1 }
+
+@[simp] lemma nat_cast_apply (n : ℕ) (m : α):
+  (n : centroid_hom α) m = n • m := rfl
+
+-- cf `add_monoid.End.semiring`
+instance : semiring (centroid_hom α) :=
+to_End_injective.semiring _ (by { ext, refl }) (by { ext, refl }) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl) (λ _ _, by { ext, refl }) (λ n, rfl)
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f,
+  { map_mul_left' := by simp [map_mul_left],
+    map_mul_right' := by simp [map_mul_right],
+    .. (-f : α →+ α) }⟩
+
+
+instance : has_sub (centroid_hom α) :=
+⟨λ f g,
+{ map_mul_left' := λ a b, by simp [map_mul_left, mul_sub],
+  map_mul_right' := λ a b, by simp [map_mul_right, sub_mul],
+  .. (f - g : α →+ α) }⟩
+
+instance has_smul_int : has_smul ℤ (centroid_hom α) :=
+⟨λ n f,
+  { map_mul_left' := λ a b,
+      by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, ←mul_smul_comm] },
+    map_mul_right' := λ a b,
+      by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, ←smul_mul_assoc] },
+    .. (n • f : α →+ α) }⟩
+
+instance : has_int_cast (centroid_hom α) :=
+{ int_cast := λ z, z • 1 }
+
+@[simp] lemma int_cast_apply (z : ℤ) (m : α):
+  (z : centroid_hom α) m = z • m := rfl
+
+instance : add_comm_group (centroid_hom α) :=
+to_End_injective.add_comm_group _ (by { ext, refl }) (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl) (λ _ _, by { ext, refl })
+
+instance : ring (centroid_hom α) := to_End_injective.ring _ (by { ext, refl }) (by { ext, refl })
+  (λ _ _, rfl)  (λ _ _, rfl) (congr_fun rfl) (congr_fun₂ rfl) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, by { ext, refl }) (λ n, rfl) (λ n, rfl)
```

Comment:
I suggest you use the `coe_` lemmas in place of `λ _ _, rfl` to make sure you didn't forget any.

Reply:
@YaelDillies Do you mean something like `(λ _ _, by rw coe_add)` ? That doesn't seem to work. Can you give me an example please?

Reply:
See for example [`set.Icc.monoid_with_zero`](https://leanprover-community.github.io/mathlib_docs/data/set/intervals/instances.html#set.Icc.monoid_with_zero)

Reply:
@YaelDillies so, should I expect to do something like this?
```
instance : ring (centroid_hom α) := to_End_injective.ring _ coe_zero coe_one
  coe_add  coe_mul coe_neg coe_sub (λ _ _, rfl) (λ _ _, rfl)
  (λ _ _, by { ext, refl }) (λ n, rfl) (λ n, rfl)
```
This doesn't quite work:
```
centroid.lean:271:0
Messages (6)
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective coe_zero
term
  coe_zero
has type
  ⇑0 = 0
but is expected to have type
  0.to_End = 0
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective _ coe_one
term
  coe_one
has type
  ⇑1 = id
but is expected to have type
  1.to_End = 1
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective _ _ coe_add
term
  coe_add
has type
  ∀ (f g : centroid_hom ?m_1), ⇑(f + g) = ⇑f + ⇑g
but is expected to have type
  ∀ (x y : centroid_hom α), (x + y).to_End = x.to_End + y.to_End
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective _ _ _ coe_mul
term
  coe_mul
has type
  ∀ (f g : centroid_hom ?m_1), ⇑(f * g) = ⇑f ∘ ⇑g
but is expected to have type
  ∀ (x y : centroid_hom α), (x * y).to_End = x.to_End * y.to_End
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective _ _ _ _ coe_neg
term
  coe_neg
has type
  ∀ (f : centroid_hom ?m_1), ⇑-f = -⇑f
but is expected to have type
  ∀ (x : centroid_hom α), (-x).to_End = -x.to_End
centroid.lean:271:52
type mismatch at application
  injective.ring to_End to_End_injective _ _ _ _ _ coe_sub
term
  coe_sub
has type
  ∀ (f g : centroid_hom ?m_1), ⇑(f - g) = ⇑f - ⇑g
but is expected to have type
  ∀ (x y : centroid_hom α), (x - y).to_End = x.to_End - y.to_End
All Messages (6)
```

Reply:
It sounds like you need `to_End_add` lemmas instead of `coe_add` then

Reply:
Okay, thanks. Done.